File: PlayerUI/ReactNative/objc/PlayerReactNative/BCOVVideoPlayer.m
//
//  BCOVVideoPlayerView.m
//  PlayerReactNative
//
//  Created by Carlos Ceja.
//

#import <AVFoundation/AVFoundation.h>
#import <AVKit/AVKit.h>

#import <BrightcovePlayerSDK/BrightcovePlayerSDK.h>

#import <React/RCTBridgeModule.h>

#import "BCOVVideoPlayer.h"


@interface BCOVVideoPlayer () <BCOVPlaybackControllerDelegate>

@end


@implementation BCOVVideoPlayer
{
    RCTEventDispatcher *_eventDispatcher;
    
    AVPlayer *_player;
    AVPlayerViewController *_avpvc;
    
    BCOVPlayerSDKManager *_manager;
    BCOVPlaybackService *_playbackService;
    id<BCOVPlaybackController> _playbackController;
}

- (instancetype)initWithEventDispatcher:(RCTEventDispatcher *)eventDispatcher
{
    if (self = [super initWithFrame:UIScreen.mainScreen.bounds])
    {
        _eventDispatcher = eventDispatcher;
        
        _player = [[AVPlayer alloc] init];
        
        _avpvc = ({
            AVPlayerViewController *avpvc = [[AVPlayerViewController alloc] init];
            avpvc.player = _player;
            avpvc.showsPlaybackControls = NO;
            avpvc;
        });
        
        _manager = [BCOVPlayerSDKManager sharedManager];
        
        [self addSubview:_avpvc.view];
    }
    
    return self;
}

- (void)setOptions:(NSDictionary *)options
{
    _options = options;
    
    NSDictionary *playbackControllerArgs = _options[@"playbackController"];
    NSNumber *autoPlay = playbackControllerArgs[@"autoPlay"];
    NSNumber *autoAdvance = playbackControllerArgs[@"autoAdvance"];
    _playbackController = ({
        id<BCOVPlaybackController> controller = [_manager createPlaybackController];
        controller.delegate = self;
        controller.autoAdvance = autoAdvance.boolValue;
        controller.autoPlay = autoPlay.boolValue;
        controller.options = @{ kBCOVAVPlayerViewControllerCompatibilityKey: @(YES) };
        controller;
    });
    
    NSDictionary *playbackServiceArgs = _options[@"playbackService"];
    NSString *accountId = playbackServiceArgs[@"accountId"];
    NSString *policyKey = playbackServiceArgs[@"policyKey"];
    _playbackService = [[BCOVPlaybackService alloc] initWithAccountId:accountId policyKey:policyKey];
    
    
    NSString *videoId = playbackServiceArgs[@"videoId"];
    NSString *authToken = ([[playbackServiceArgs[@"authToken"] class] isKindOfClass:[NSNull class]] ?
                           playbackServiceArgs[@"authToken"] :
                           nil);
    NSDictionary *parameters = ([[playbackServiceArgs[@"parameters"] class] isKindOfClass:[NSNull class]] ?
                                playbackServiceArgs[@"parameters"] :
                                nil);
    
    [_playbackService findVideoWithVideoID:videoId authToken:authToken parameters:parameters completion:^(BCOVVideo *video, NSDictionary *jsonResponse, NSError *error)
     {
        if (video)
        {
            [self->_playbackController setVideos:@[ video ]];
        }
        
    }];
}

- (void)playPause:(BOOL)isPlaying
{
    if (isPlaying)
    {
        [_player pause];
    }
    else
    {
        [_player play];
    }
}

#pragma mark BCOVPlaybackControllerDelegate

- (void)playbackController:(id<BCOVPlaybackController>)controller didAdvanceToPlaybackSession:(id<BCOVPlaybackSession>)session
{
    _player = session.player;
    _avpvc.player = session.player;
    
    if (self.onReady)
    {
        id duration = session.video.properties[@"duration"];
        self.onReady(@{@"duration": duration});
    }
}

- (void)playbackController:(id<BCOVPlaybackController>)controller playbackSession:(id<BCOVPlaybackSession>)session didProgressTo:(NSTimeInterval)progress
{
    if (self.onProgress && !isinf(progress))
    {
        self.onProgress(@{@"progress": @(progress)});
    }
}

@end

File: PlayerUI/ReactNative/swift/PlayerReactNative/BCOVVideoPlayer.swift
//
//  BCOVVideoPlayer.swift
//  PlayerReactNative
//
//  Created by Carlos Ceja.
//

import AVFoundation
import AVKit
import Foundation
import UIKit

import BrightcovePlayerSDK
import React

class BCOVVideoPlayer: UIView {
    
    @objc var options: NSDictionary? {
        didSet {
            let parsedData = options as! [String: Any]
            
            let playbackControllerArgs = parsedData["playbackController"] as! [String: Any]
            let autoPlay = playbackControllerArgs["autoPlay"] as! Bool
            let autoAdvance = playbackControllerArgs["autoAdvance"] as! Bool
            
            self.playbackController = self.manager.createPlaybackController()!
            self.playbackController.delegate = self
            self.playbackController.isAutoPlay = autoPlay
            self.playbackController.isAutoAdvance = autoAdvance
            self.playbackController.options = [kBCOVAVPlayerViewControllerCompatibilityKey: true]
            
            let playbackServiceArgs = parsedData["playbackService"] as! [String: Any]
            let accountId = playbackServiceArgs["accountId"] as! String
            let policyKey = playbackServiceArgs["policyKey"] as! String
            self.playbackService = BCOVPlaybackService(accountId: accountId, policyKey: policyKey)
            
            let videoId = playbackServiceArgs["videoId"] as! String
            let authToken = playbackServiceArgs["authToken"] as? String
            let parameters = playbackServiceArgs["parameters"] as? [AnyHashable : Any]
            
            var configuration = [kBCOVPlaybackServiceConfigurationKeyAssetID:videoId]
            if authToken != nil {
                configuration[kBCOVPlaybackServiceConfigurationKeyAuthToken] = authToken
            }
            playbackService?.findVideo(withConfiguration: configuration, queryParameters: parameters, completion: { [weak self] (video: BCOVVideo?, jsonResponse: [AnyHashable: Any]?, error: Error?) in
                if let video = video {
                    self?.playbackController.setVideos([video] as NSFastEnumeration)
                }
            })
        }
    }
    
    @objc var onReady: RCTDirectEventBlock?
    
    @objc var onProgress: RCTDirectEventBlock?
    
    @objc var eventDispatcher: RCTEventDispatcher!
    
    lazy private var player: AVPlayer = {
        var _player = AVPlayer()
        return _player
    }()
    
    lazy private var avpvc: AVPlayerViewController = {
        var _avpvc = AVPlayerViewController()
        _avpvc.player = self.player
        _avpvc.showsPlaybackControls = false
        return _avpvc
    }()
    
    lazy private var manager: BCOVPlayerSDKManager = {
        let _manager = BCOVPlayerSDKManager.shared()!
        return _manager
    }()
    
    private var playbackService: BCOVPlaybackService!
    private var playbackController: BCOVPlaybackController!
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        self.addSubview(self.avpvc.view)
    }
    
    func setEventDispatcher(_ eventDispatcher: RCTEventDispatcher) {
        self.eventDispatcher = eventDispatcher
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    @objc func playPause(_ isPlaying: Bool) {
        if isPlaying {
            self.player.pause()
        }
        else {
            self.player.play()
        }
    }
}


// MARK: - BCOVPlaybackControllerDelegate

extension BCOVVideoPlayer: BCOVPlaybackControllerDelegate {
    
    func playbackController(_ controller: BCOVPlaybackController!, didAdvanceTo session: BCOVPlaybackSession!) {
        
        self.player = session.player
        self.avpvc.player = session.player
        
        if (self.onReady != nil) {
            let duration = session?.video.properties["duration"] as! NSNumber
            self.onReady?(["duration": duration])
        }
    }
    
    func playbackController(_ controller: BCOVPlaybackController!, playbackSession session: BCOVPlaybackSession!, didProgressTo progress: TimeInterval) {
        
        if (self.onProgress != nil && !progress.isInfinite) {
            self.onProgress?(["progress": NSNumber(value: progress)])
        }
    }
}
